From 303dfac81b5fac0203a61fea6a73f31e9f36b0c7 Mon Sep 17 00:00:00 2001
From: Rajat Soni <quic_rajson@quicinc.com>
Date: Fri, 29 Mar 2024 18:17:56 +0530
Subject: [PATCH] hostapd: Fix channel switch to DFS channel

When we are configuring option channel auto, we are not able to switch to
given dfs channel because when we are trying to move to a dfs channel
interface is disabled and enabled again. When interface is disabled and
enabled we are setting iface's freq and channel to 0 in function
setup_interface2() in case of acs is enabled, now we don't know which
channel we were trying to move. Now ACS will run and 5G will be
up in the channel that is suitable.

To fix this issue add flag named is_ch_switch_dfs to check if channel
switch request for dfs channel and we can use this in setup_interface2()
to decide that we have to set iface's freq and channel to 0 or not.
This way iface's freq and channel will retain the values while channel
switch to dfs channel when acs is enable.

Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
---
Note:
Sent for internal review.
Subject: hostapd: Fix channel switch to DFS channel
---
 hostapd/ctrl_iface.c | 1 +
 src/ap/hostapd.c     | 4 +++-
 src/ap/hostapd.h     | 2 ++
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 1fecd20..ba8d383 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -3017,6 +3017,7 @@ static int hostapd_ctrl_iface_chan_switch(struct hostapd_iface *iface,
 			   settings.freq_params.center_freq1);
 
 		/* Perform CAC and switch channel */
+		iface->is_ch_switch_dfs = true;
 		hostapd_switch_channel_fallback(iface, &settings.freq_params);
 		return 0;
 	}
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 951d018..239c963 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -2194,10 +2194,11 @@ static int setup_interface2(struct hostapd_iface *iface)
 	} else {
 		int ret;
 
-		if (iface->conf->acs) {
+		if (iface->conf->acs && !iface->is_ch_switch_dfs) {
 			iface->freq = 0;
 			iface->conf->channel = 0;
 		}
+		iface->is_ch_switch_dfs = false;
 
 		ret = configured_fixed_chan_to_freq(iface);
 		if (ret < 0)
@@ -3272,6 +3273,7 @@ struct hostapd_iface * hostapd_init(struct hapd_interfaces *interfaces,
 		hostapd_bss_setup_multi_link(hapd, interfaces);
 #endif /* CONFIG_IEEE80211BE */
 	}
+	hapd_iface->is_ch_switch_dfs = false;
 
 	return hapd_iface;
 
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index f69a8b3..8b0057a 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -785,6 +785,8 @@ struct hostapd_iface {
 	struct hostapd_multi_hw_info *multi_hw_info;
 	u8 num_multi_hws;
 	struct hostapd_multi_hw_info *current_hw_info;
+	/* Channel switch from ACS to DFS*/
+	bool is_ch_switch_dfs;
 };
 
 /* Iterate over all affiliated links in MLD to which @self belongs.
-- 
2.34.1

